application/common/models/Password.php

```diff
@@ -214,5 +222,64 @@ public function save()
         }
     }
 
+    public function validateNotPublicPassword($attribute)
+    {
+        /*
+         * block passwords provided in https://www.youtube.com/watch?v=3LRnjzA8auk&list=PLu5OsENIeX656zXJ96FCL169WNmvnPveo
+         */
+        $publicPasswords = [
+            'one4amzn',
+            'one4ggle',
+            'one4ebay',
+            '$$Ymh7Hp3dfgQr9L#!>s;',
+            '7startpenguins!snap',
+            'deserty.domes.slide2',
+            'about-slithers-quakely.',
+        ];
 
+        foreach ($publicPasswords as $publicPassword) {
+            if ($this->$attribute == $publicPassword) {
+                $msg = "Ha ha, good one. No, you shouldn't use a password you saw " .
+                    'in a video about creating good passwords.';
+                $this->addError($attribute, \Yii::t('app', $msg));
+            }
+        }
+    }
+
+    /**
+     * Validate a password against the havibeenpwned security breach database.
+     *
+     * Inspired by https://github.com/PayFast/laravel-validate-pwned-password
+     *
+     * @param $attribute
+     */
+    public function validateNotPwned($attribute)
+    {
+        $hash = sha1($this->$attribute);
+        $hashPrefix = substr($hash, 0, 5);
+        $url = 'https://api.pwnedpasswords.com/range/' . $hashPrefix;
```

Comment:
It would probably be a good idea to [specify what version of the API we want](https://haveibeenpwned.com/API/v2#APIVersion), so that future (possibly breaking?) changes in new versions won't affect us until we've had time to refactor to support the new API version.